src/value.rs

```diff
@@ -293,6 +294,36 @@ impl Value {
 
         (bytes, bit_length)
     }
+
+    /// Check if the value is of the given type.
+    pub fn is_of_type(self: Arc<Self>, ty: Arc<Final>) -> bool {
```

Comment:
In d9ca9252c58e16c6375d4d570020d77b6b3c96a6:

I'm fairly sure we could take an `&Arc<Self>` and `&Arc<Final>` here and save a bunch of clones. We would also want `split_*`/`unwrap_*` to return references.

Personally I would suggest having a family `to_sum`/`to_left`/etc which returns `Arc<Self>` as they do now, and also a family `as_sum`/`as_left`/etc which return `&Self`. I think you can then use the `as_` methods, have your stack be holding references rather than Arcs, and save yourself a lot of atomic refcounts.

Reply:
I've been wrestling with the question `Arc<Self>` vs `&Self` for a while. Sometimes we need an owned `Arc<Self>`. Sometimes we want to avoid unnecessary clones using `&Self`. Supporting both sounds like a good compromise.